crates/re_viewport/src/space_view_heuristics.rs

```diff
@@ -499,37 +499,59 @@ pub fn identify_entities_per_system_per_class(
-                        continue;
-                    };
-
-                    for system in systems {
-                        // TODO(andreas/jleibs): This is only needed because of images.
-                        // The `queries_any_components_of` method should go away entirely after #3032 lands
-                        if let Ok(view_part_system) = part_collection.get_by_name(*system) {
-                            if !view_part_system.queries_any_components_of(
-                                store,
-                                ent_path,
-                                &components,
-                            ) {
-                                continue;
-                            }
-                        }
+        let all_components: IntSet<_> = components.into_iter().collect();
+
+        for (required_components, systems_per_class) in &systems_per_required_components {
+            if !all_components.is_superset(&required_components.0) {
+                continue;
+            }
 
-                        per_class_per_system_entities
-                            .entry(*class)
-                            .or_default()
-                            .entry(*system)
-                            .or_default()
-                            .insert(ent_path.clone());
+            for (class, systems) in systems_per_class {
+                let Some((_, part_collection)) = system_collections_per_class.get(class) else {
+                    continue;
+                };
+
+                for system in systems {
+                    if let Ok(view_part_system) = part_collection.get_by_name(*system) {
+                        if !view_part_system.heuristic_filter(store, ent_path, &all_components) {
+                            continue;
+                        }
                     }
+
+                    entities_per_system_per_class
+                        .entry(*class)
+                        .or_default()
+                        .entry(*system)
+                        .or_default()
+                        .insert(ent_path.clone());
                 }
             }
         }
     }
 
-    per_class_per_system_entities
+    entities_per_system_per_class
+}
+
+// ---
+
+#[derive(Debug, Clone, Eq)]
+struct HashableIntSet<T: nohash_hasher::IsEnabled + std::hash::Hash>(IntSet<T>);
+
+impl<T: nohash_hasher::IsEnabled + std::hash::Hash + PartialEq> PartialEq for HashableIntSet<T> {
+    fn eq(&self, other: &Self) -> bool {
+        self.0.len() == other.0.len()
+            && self
+                .0
+                .iter()
+                .zip(other.0.iter())
+                .all(|(left, right)| left == right)
+    }
+}
+
+impl<T: nohash_hasher::IsEnabled + std::hash::Hash> std::hash::Hash for HashableIntSet<T> {
+    fn hash<H: std::hash::Hasher>(&self, state: &mut H) {
+        for value in &self.0 {
+            value.hash(state);
+        }
+    }
```

Comment:
It seems I just went and assumed that `IntSet` is a btree for some reason, except it probably is not :grimacing:

Reply:
Oh nevermind, it's nohash'd so it should in fact have a stable order then... hmmm :thinking:

Reply:
This does not look good to me.

Two hashsets can have different iteration orders depending on insertion order, even when they contain the exact same elements

Reply:
If you really, really, want to do this, then xor the hashes of the elements together and return the result as hash. xor is order-independent.

then it can even implement `nohash_hasher::IsEnabled`